src/cpu/flexcpu/simple_dataflow_cpu.hh

```diff
@@ -0,0 +1,649 @@
+
+        void sendPacket(PacketPtr pkt);
+
+        bool blocked() { return blockedPkt != nullptr; }
+    };
+
+    /**
+     * Very simple wrapper for the Translation interface, which takes a
+     * std::function in any form. This allows the use of lambdas to handle the
+     * completion of the translation instead of needing specialized polymorphic
+     * classes.
+     *
+     * Note that instances of this class are assumed to be allocated through
+     * Operator::new, because they will self-delete upon calling finish(). Note
+     * that this also requires the assumption that all CallbackTransHandlers
+     * constructed will have finish() called on them, otherwise this will cause
+     * a memory leak.
+     */
+    class CallbackTransHandler : public BaseTLB::Translation
+    {
+      protected:
+        SimpleDataflowCPU *cpu;
+        TranslationCallback callback;
+        Tick sendTime;
+      public:
+        CallbackTransHandler(SimpleDataflowCPU *cpu,
+                             TranslationCallback callback,
+                             Tick send_time):
+            cpu(cpu), callback(callback), sendTime(send_time)
+        { }
+
+        void markDelayed() override
+        { }
+
+        void finish(const Fault& fault, const RequestPtr& req,
+                    ThreadContext* tc, BaseTLB::Mode mode) override
+        {
+            callback(fault, req);
+
+            // It requires assumptions that users of this will be well-behaved
+            // to call "delete this;" in a C++ class, but since this is an
+            // internal class, we should be able to make this assumption.
+            delete this;
+        }
+    };
+
+    // END Internal class definitions
+
+    // BEGIN Internal container structs
+
+    struct SplitAccCtrlBlk
+    {
+        // Packet used to communicate to the instruction (completeAcc)
+        // The data always is allocated in this packet. The other packets have
+        // static pointers to the data here.
+        PacketPtr main = nullptr;
+
+        // Used for sending/receiving data from memory system
+        PacketPtr low = nullptr;
+        bool lowReceived = false;
+        PacketPtr high = nullptr;
+        bool highReceived = false;
+    };
+
+    // END Internal container structs
+
+
+    // BEGIN Internal parameters
+
+    Cycles executionLatency;
+
+    // END Internal parameters
+
+    /**
+     * Models a functional unit. Each unit has a latency, in cycles, and a
+     * bandwidth in accesses per cycle. A latency of 0 implies the action is
+     * immediately taken. Otherwise, the action of the functional unit is not
+     * functionally "executed" until a clock edge after latency cycles.
+     *
+     * NOTE: We might want to make this a SimObject so it can be configured
+     *       from python more easily.
+     */
+    class Resource
+    {
+      protected:
+        SimpleDataflowCPU* cpu;
+
+      private:
+        const Cycles latency;
+        const int bandwidth;
```

Comment:
Could you use the MemoryBandwidth type in gem5 for this?

Reply:
Different semantic